File: mobile/lib/core/models/round.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// 라운드 모델 (Firestore rounds/{roundId})
class Round {
  final String id;
  final String courseId;
  final String courseName;
  final String region;
  final int holesCount;
  final String createdBy;
  final String status; // ACTIVE, FINISHED
  final DateTime? date;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? teeTime;
  final String? tee;
  final String? backCourseId;
  final String? backCourseName;
  /// 라운드 번호 4자리 (등록 시 랜덤 생성)
  final String? roundNo;
  /// 골프장명 (목록 상단 표시용)
  final String? golfCourseName;

  const Round({
    required this.id,
    required this.courseId,
    required this.courseName,
    required this.region,
    this.holesCount = 18,
    required this.createdBy,
    this.status = 'ACTIVE',
    this.date,
    this.createdAt,
    this.updatedAt,
    this.teeTime,
    this.tee,
    this.backCourseId,
    this.backCourseName,
    this.roundNo,
    this.golfCourseName,
  });

  factory Round.fromMap(String id, Map<String, dynamic> map) {
    return Round(
      id: id,
      courseId: map['courseId'] as String? ?? '',
      courseName: map['courseName'] as String? ?? '',
      region: map['region'] as String? ?? '',
      holesCount: (map['holesCount'] as int?) ?? 18,
      createdBy: map['createdBy'] as String? ?? '',
      status: map['status'] as String? ?? 'ACTIVE',
      date: _toDate(map['date']),
      createdAt: _toDate(map['createdAt']),
      updatedAt: _toDate(map['updatedAt']),
      teeTime: map['teeTime'] as String?,
      tee: map['tee'] as String?,
      backCourseId: map['backCourseId'] as String?,
      backCourseName: map['backCourseName'] as String?,
      roundNo: map['roundNo'] as String?,
      golfCourseName: map['golfCourseName'] as String?,
    );
  }

  static DateTime? _toDate(dynamic v) {
    if (v == null) return null;
    if (v is Timestamp) return v.toDate();
    return v as DateTime;
  }

  bool get isActive => status == 'ACTIVE';
}

File: mobile-rn/src/core/types/round.ts
/**
 * 라운드 문서 (Firestore rounds/{roundId})
 * 규칙: createdBy로 HOST 판별, status로 종료 여부 판별
 */
export type RoundStatus = 'DRAFT' | 'IN_PROGRESS' | 'FINISHED';

export interface Round {
  id: string;
  createdBy: string;
  roundName: string | null;
  roundNumber: string | null; // 4자리
  golfCourseId: string;
  golfCourseName: string;
  frontCourseId: string;
  frontCourseName: string;
  backCourseId: string;
  backCourseName: string;
  courseId: string; // 하위 호환: 전반 코스
  courseName: string;
  teeTime: string | null;
  status: RoundStatus;
  scheduledAt: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

/**
 * 라운드 참가자 (rounds/{roundId}/participants/{uid})
 */
export interface RoundParticipant {
  uid: string;
  nickname: string | null;
  role: 'HOST' | 'MEMBER';
  joinStatus: string;
  holesEntered: number;
  totalOut: number;
  totalIn: number;
  total: number;
  updatedAt: Date;
  /** 스코어 확정 시각. 있으면 확정 완료 상태 */
  scoreConfirmedAt: Date | null;

  // ---- HomeScreen 통계 최적화를 위한 집계값 (옵션/하위호환) ----
  // 기존 데이터(새 필드 없음)는 fetchUserConfirmedRoundStats에서 fallback으로 계산합니다.
  totalPutts?: number; // 18홀 퍼팅 합
  girHitCount?: number; // strokes-putts <= 2 인 홀 수
  girTotalCount?: number; // GIR 계산 분모(통상 18)
  firHitCount?: number; // par=4/5 홀 중 fairway=true인 홀 수
  firTotalCount?: number; // par=4/5 홀 수
}

/** 홀별 스코어 (rounds/{roundId}/scores/{uid} holes.{n}) */
export interface HoleScoreData {
  strokes: number;
  putts: number;
  fairway?: boolean | null;
  rough?: boolean | null;
  penalty?: boolean | null;
  ob?: number | null;
}
